### ISSUE TITLE
Drawing Index of stock chart

### ISSUE THREAD

**Issue by workingbuddy10** [amcharts/amcharts5#1327]
In the recent update 5.8.0 there are some changes in the drawing index behavior.

Firstly can you please tell me how it working now? As I can see the index get updated on repeating the drawing. Still would like to know the whole logic behind the index initialization.

Earlier If I wanted to delete the drawings I used to do this :
```
var lineSeries;
mainPanel.drawings.each(function(series) {
  if (series.className == "SimpleLineSeries") {
    lineSeries = series;
  }
})
lineSeries.disposeIndex(1);
```

Now how I can achieve something similar functionality. Like I want to retrieve the un-serialize drawings only earlier I have given the pre-default index which have I stored in a map so I can easily delete the drawings.

I was doing like this : 
```
    const chartInstances = this.chartStore.chartInstances;
    chartInstances.forEach((chartInstance) => {
      const drawingData = chartInstance.mainPanel.drawings._values;
      drawingData.forEach((series) => {
        if (VALID_SERIES_TYPES.includes(series.className)) {
          const index = AUTO_INDICATOR_INDEX_MAP[indicator];
          if (index) {
            console.log(index);
            console.log(series._index);  
            series._disposeIndex(index);
          }
        }
      });
      
    });
```
Using un-serialize drawings : Now If I draw a polyline & delete it, then again draw the polyline & try to delete it the index I am getting from `(series_index)` is different.

Before un-serializing drawings I have given the index of polyline as 900, so when I try to delete it for the first time I am getting the 900 index only so it is deleting perfectly, but when I again use the polyline I get the index as 1800. Which violates my condition in the code as it is looking for the index 900 which is given in the `AUTO_INDICATOR_INDEX_MAP`.

So my main goal is to delete the specific un-serialize drawings only, earlier it was getting fulfilled as the index was not changing in while  performing un-serialize drawings ( also with the help of  AUTO_INDICATOR_INDEX_MAP which I have stored) but now it is changing as in the code the` this._updateDrawingIndexes` is being called before  un-serializing drawings.

**Comment by martynasma** [comment]
The change introduced in 5.8.0 was that drawing unserializer would check whether there are user-added drawings at the moment of unserialization, and would shift indexes of the drawings being unserialized so that they start at the next index than the latest user-created drawing.

This change was implemented based on your earlier requests.

If there are no user-created drawings, no changes should be applied to unserialized drawings indexes.

I understand that this makes identifying of precise drawings problematic.

We will discuss ways to add some kind of a reliable drawing identification method. Will let you know.

**Comment by workingbuddy10** [comment]
Thanks @martynasma for considering my request!!

1. One suggestion is that we can add a field in the drawing data template _drawingType : "User Drawn/unserialized/local storage. And there will be `get drawingInfo` function which can helps to retrieve the index & the type of drawings.
2. Or we can provide the alpha- numeric index to the drawings.
3. Or just don't Adjust/change the index of unserialized  drawings. Adjust the index of user made & local storage drawings, as we only have to target the unserialized  drawings. Rest user made & local storage drawings are same & can be deleted by eraser/delete button!

I believe going with option 2 or 3 would be better. Rest keen to know your suggestion & approach.

**Comment by martynasma** [comment]
> Or just don't Adjust/change the index of unserialized drawings. Adjust the index of user made & local storage drawings, as we only have to target the unserialized drawings. Rest user made & local storage drawings are same & can be deleted by eraser/delete button!

This is not an option, I'm afraid. Changing index of an active drawing would be super complicated due to how they are organized internally.

Easier to do that in the source JSON before import.

I'm thinking maybe we can introduce some kind of additional `id` value for drawings, which would use some kind of pseudo-unique logic to auto-generate those ids. Like index + timestamp, or similar.

Then have a similar method to `disposeIndex()` but which would take id as a parameter instead.

We'll discuss this.

**Comment by workingbuddy10** [comment]
> I'm thinking maybe we can introduce some kind of additional `id` value for drawings, which would use some kind of pseudo-unique logic to auto-generate those ids. Like index + timestamp, or similar.
> 
> Then have a similar method to `disposeIndex()` but which would take id as a parameter instead.

Then why don't you make id as alpha-numeric

> 2. Or we can provide the alpha- numeric index to the drawings.


By this we don't have any limitation to the timestamp (Like index + timestamp)


It's just I need to know before hand & it should be fixed only that what id is there in the un-serialized drawings so that I can have the map of the index in my frontend and can easily target & remove by `disposeIndex()`

**Comment by martynasma** [comment]
Alphanumeric index is hard to auto-increment.

**Comment by workingbuddy10** [comment]
> Alphanumeric index is hard to auto-increment.

True, I can agree on this!

> It's just I need to know before hand only that what id is there in the un-serialized drawings so that I can have the map of the index in my frontend and can easily target & remove by `disposeIndex()`

**Comment by workingbuddy10** [comment]
@martynasma Till that time can you please remove the changes which you did in  5.8.0 related to drawing index?

**Comment by martynasma** [comment]
No, since it wasn't right. It could lead to duplicate indexes.

You may consider using some earlier version.

**Comment by workingbuddy10** [comment]
Yes I am doing that only, but I won't be getting the new upgrades/fixes. ( Though most of things are fine if I don't upgrade) 
Just checking if it is possible.

**Comment by workingbuddy10** [comment]
@martynasma, what is decided regarding the index?
What will be the whole logic, you are considering?

**Comment by martynasma** [comment]
Here's a snippet of the upcoming changelog. Should provide details:

```
### Added
- New `DrawingSeries` method: `getIndex(drawingId)`. Will return index of a drawing corresponding to a unique id, or `null` if not found.

### Changed
- All `DrawingSeries` will now add a unique `drawingId` to all drawing data.
```

**Comment by workingbuddy10** [comment]
What happens in the case of the  un-serialize drawings?
Does the index  of the  un-serialize drawings changes or it remains the same?

@martynasma

**Comment by martynasma** [comment]
The index can change. ID will stay the same. You can use `getIndex()` method to convert static id to a current index.

**Comment by workingbuddy10** [comment]
> So my main goal is to delete the specific un-serialize drawings only, earlier it was getting fulfilled as the index was not changing in while performing un-serialize drawings ( also with the help of AUTO_INDICATOR_INDEX_MAP which I have stored) but now it is changing as in the code the` this._updateDrawingIndexes` is being called before un-serializing drawings.

So now instead of storing the index of the specific  un-serialize drawings I can save the ID in my MAP.

Later if I want to delete the specific  un-serialize drawing I can just retrieve the index via `getIndex(drawingId)` and then dispose it.

Am I right? @martynasma

**Comment by martynasma** [comment]
Correct.

**Comment by workingbuddy10** [comment]
Just one small doubt @martynasma , what is the typeof ID.
Alpha -numeric or string ?

**Comment by workingbuddy10** [comment]
I have a given a drop-down where I have provided the option to user to check/un-check the  un-serialize drawing names. 
So when user checks the  drawing `A` then it  un-serialize drawings and when user un-checks then it retrieves the index and disposes/delete it.

In the case of Delete All Drawings (Drawing toolbar) I didn't wanted to delete the drop -down/ un-serialize drawings so I used to check the index of the drawings if it is less than 500 then I used to assume it as a user-drawn drawings so using Delete All Drawings button delete all the user-drawn drawings Only as  un-serialize drawings have indexes more than 1000.

Like this : 
```
  deleteAllDrawings = () => {
    Object.values(this.selectedDrawingControl._drawingSeries).forEach(
      (seriesList) => {
        seriesList.forEach((seriesItem) => {
          if (seriesItem._index < 500) {
            seriesItem.clearDrawings();
          }
        });
      }
    );
    this.disableDrawingTools();
  };
```

Now how will I achieve the same functionality (Delete All Drawings don't delete  un-serialize drawings)? Do I have to check the ID or what?
@martynasma

**Comment by martynasma** [comment]
> Just one small doubt @martynasma , what is the typeof ID. Alpha -numeric or string ?

ID can be any string.

**Comment by martynasma** [comment]
Implemented in [5.8.4](https://github.com/amcharts/amcharts5/releases/tag/5.8.4).

## [5.8.4] - 2024-02-22

### Added
- New `DrawingSeries` method: `getIndex(drawingId)`. Will return index of a drawing corresponding to a unique id, or `null` if not found.

### Changed
- `PeriodSelector` will now zoom to last/first data items if `stockSeries` is set on `StockChart`. Previously it used to zoom to X-axis scale end/start, regardless of data.
- All `DrawingSeries` will now add a unique `drawingId` to all drawing data.
- Stock indicatos adjusted so that they could handle data with negative values, `0` volume values and all OHLC values being equal.

### Fixed
- Scale of a `ValueAxis` was not adjusting properly if its height/width changed during zoom.
- Standalone `DrawingControl` (one that was not added to a toolbar) was erroring out when parent `Root` was disposed.
- HTML content in a `Label` (with `html` set) was not inheriting labels actual color (`fill`).
- Setting a `selectedDataItem` to a data item without children on a `Hierarchy` series was resulting in error.
- Heat rules were setting "neutral" color on the default state of a target even if it was undefined, resulting in some issues when value was not set.
- Heat rules were ignoring `customFunction` if value was not set.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
